onnxruntime/contrib_ops/cpu/non_max_suppression.cc

```diff
@@ -37,119 +35,183 @@ void NonMaxSuppression<T>::MaxMin(const T& lhs, const T& rhs, T& min, T& max) co
+  const float area2 = (x2_max - x2_min) * (y2_max - y2_min);
+  const float union_area = area1 + area2 - intersection_area;
 
-  if (area1 <= static_cast<T>(0.0) || area2 <= static_cast<T>(0.0) || union_area <= static_cast<T>(0.0)) {
+  if (area1 <= static_cast<float>(0.0) || area2 <= static_cast<float>(0.0) || union_area <= static_cast<float>(0.0)) {
     return false;
   }
 
-  const T intersection_over_union = intersection_area / union_area;
+  const float intersection_over_union = intersection_area / union_area;
 
   return intersection_over_union > iou_threshold_;
 }
 
-template <typename T>
-Status NonMaxSuppression<T>::Compute(OpKernelContext* ctx) const {
+Status NonMaxSuppression::ParepareCompute(const TensorShape& boxes_shape, const TensorShape& scores_shape) const {
+  ORT_RETURN_IF_NOT(boxes_shape.NumDimensions() == 4, "boxes must be a 4D tensor.");
+  ORT_RETURN_IF_NOT(scores_shape.NumDimensions() == 3, "scores must be a 3D tensor.");
+
+  auto boxes_dims = boxes_shape.GetDims();
+  auto scores_dims = scores_shape.GetDims();
+  ORT_RETURN_IF_NOT(boxes_dims[0] == scores_dims[0], "boxes and scores should have same num_batches.");
+  ORT_RETURN_IF_NOT(boxes_dims[1] == scores_dims[1], "boxes and scores should have same num_classes.");
+  ORT_RETURN_IF_NOT(boxes_dims[2] == scores_dims[2], "boxes and scores should have same spatial_dimention.");
+  ORT_RETURN_IF_NOT(boxes_dims[3] == 4, "The most inner dimension in boxes must have 4 data.");
+
+  const_cast<int64_t&>(num_batches_) = boxes_dims[0];
+  const_cast<int64_t&>(num_classes_) = boxes_dims[1];
+  const_cast<int64_t&>(num_boxes_) = boxes_dims[2];
+
+  return Status::OK();
+}
+
+Status NonMaxSuppression::Compute(OpKernelContext* ctx) const {
   const Tensor* boxes = ctx->Input<Tensor>(0);
   ORT_ENFORCE(boxes);
   const Tensor* scores = ctx->Input<Tensor>(1);
   ORT_ENFORCE(scores);
 
-  const TensorShape& boxes_shape = boxes->Shape();
-  auto boxes_dims = boxes_shape.GetDims();
-  ORT_RETURN_IF_NOT(boxes_shape.NumDimensions() == 2, "boxes must be a 2D tensor.");
-  int64_t num_boxes = boxes_dims[0];
-  ORT_RETURN_IF_NOT(boxes_dims[1] == 4, "boxes shape must be a 2D tensor with shape [num_boxes, 4].");
-
-  const TensorShape& scores_shape = scores->Shape();
-  ORT_RETURN_IF_NOT(scores_shape.NumDimensions() == 1, "boxes must be a 1D tensor.");
-  ORT_RETURN_IF_NOT(scores_shape.GetDims()[0] == num_boxes, "scores and boxes should have same num_boxes.");
-
-  if (max_output_size_ <= 0 || boxes_dims[0] == 0) {
-    TensorShape output_shape({0});
-    ctx->Output(0, output_shape);
-    return Status::OK();
-  }
+  auto& boxes_shape = boxes->Shape();
+  auto& scores_shape = scores->Shape();
+  auto ret = ParepareCompute(boxes_shape, scores_shape);
+  ORT_RETURN_IF_NOT(ret.IsOK(), ret.ErrorMessage());
 
-  const T* boxes_data = boxes->Data<T>();
-  const T* scores_data = scores->Data<T>();
+  const float* boxes_data = boxes->Data<float>();
+  const float* scores_data = scores->Data<float>();
```

Comment:
Could scores be any type other that float? it comes from Tensor and most of the time is "float" but could be "float16" at one point

Reply:
cpu don't have float 16 kernel. If it is float16, a extra cast operator will be added to cast the data to float, then cast back to float16 for output.